counter/src/main/java/org/infinispan/counter/impl/strong/AbstractStrongCounter.java

```diff
@@ -0,0 +1,187 @@
+public abstract class AbstractStrongCounter implements StrongCounter {
+
+   private static final AtomicReferenceFieldUpdater<AbstractStrongCounter, CounterValue> WEAK_COUNTER_UPDATER =
+         newUpdater(AbstractStrongCounter.class, CounterValue.class, "weakCounter");
+
+   final StrongCounterKey key;
+   private final FunctionalMap.ReadWriteMap<StrongCounterKey, CounterValue> readWriteMap;
+   private final FunctionalMap.ReadOnlyMap<StrongCounterKey, CounterValue> readOnlyMap;
+   private final NotificationManager notificationManager;
+   private volatile CounterValue weakCounter;
+
+   AbstractStrongCounter(String counterName, AdvancedCache<StrongCounterKey, CounterValue> cache, Storage storage) {
+      FunctionalMapImpl<StrongCounterKey, CounterValue> functionalMap = FunctionalMapImpl.create(cache)
+            .withParams(Utils.getPersistenceMode(storage));
+      this.key = new StrongCounterKey(counterName);
+      this.readWriteMap = ReadWriteMapImpl.create(functionalMap);
+      this.readOnlyMap = ReadOnlyMapImpl.create(functionalMap);
+      this.notificationManager = new NotificationManager();
+      this.weakCounter = null;
+      registerListener(cache);
+   }
+
+   public final void init(CounterConfiguration configuration) {
+      try {
+         CounterValue existingValue = readWriteMap.eval(key, new InitializeCounterFunction<>(configuration)).get();
+         initCounterState(existingValue == null ? newCounterValue(configuration) : existingValue);
+      } catch (InterruptedException e) {
+         Thread.currentThread().interrupt();
+         throw new CounterException(e);
+      } catch (ExecutionException e) {
+         throw Utils.rethrowAsCounterException(e);
+      }
+   }
+
+   @Override
+   public final String getName() {
+      return key.getCounterName().toString();
+   }
+
+   @Override
+   public final CompletableFuture<Long> getValue() {
+      return readOnlyMap.eval(key, ReadFunction.getInstance()).thenApply(this::handleReadResult);
+   }
+
+   @Override
+   public final long weakGetValue() {
+      return weakCounter.getValue();
+   }
+
+   @Override
+   public final CompletableFuture<Long> addAndGet(long delta) {
+      return readWriteMap.eval(key, new AddFunction<>(delta)).thenApply(this::handleAddResult);
+   }
+
+   @Override
+   public final CompletableFuture<Void> reset() {
+      return readWriteMap.eval(key, ResetFunction.getInstance());
+   }
+
+   @Override
+   public final <T extends CounterListener> Handler<T> addListener(T listener) {
+      return notificationManager.addListener(listener);
+   }
+
+   @Override
+   public CompletableFuture<Boolean> compareAndSet(long expect, long update) {
+      return readWriteMap.eval(key, new CompareAndSetFunction<>(expect, update)).thenApply(this::handleCASResult);
+   }
+
+   protected abstract Boolean handleCASResult(CounterState state);
+
+   @CacheEntryModified
+   public void updateState(CacheEntryEvent<StrongCounterKey, CounterValue> event) {
```

Comment:
I don't think it's a right thing to have a compulsory weak value on strong counters, because the listeners here bring some overhead. When the user wants a strong counter, let's give him exactly that. When he wants eventually updated counter, he should use the weak one.